pkg/hub/csr/reconciler.go

```diff
@@ -0,0 +1,228 @@
+	"k8s.io/klog/v2"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+	"open-cluster-management.io/registration/pkg/hub/user"
+)
+
+type reconcileState int64
+
+const (
+	reconcileStop reconcileState = iota
+	reconcileContinue
+)
+
+type csrInfo struct {
+	name       string
+	labels     map[string]string
+	signerName string
+	username   string
+	uid        string
+	groups     []string
+	extra      map[string]authorizationv1.ExtraValue
+	request    []byte
+}
+
+type approveCSRFunc func(kubernetes.Interface) error
+
+type Reconciler interface {
+	Reconcile(context.Context, csrInfo, approveCSRFunc) (reconcileState, error)
+}
+
+type csrRenewalReconciler struct {
+	kubeClient    kubernetes.Interface
+	eventRecorder events.Recorder
+}
+
+func NewCSRRenewalReconciler(kubeClient kubernetes.Interface, recorder events.Recorder) Reconciler {
+	return &csrRenewalReconciler{
+		kubeClient:    kubeClient,
+		eventRecorder: recorder.WithComponentSuffix("csr-approving-controller"),
+	}
+}
+
+func (r *csrRenewalReconciler) Reconcile(ctx context.Context, csr csrInfo, approveCSR approveCSRFunc) (reconcileState, error) {
+	// Check whether current csr is a valid spoker cluster csr.
+	valid, commonName := validateCSR(csr)
+	if !valid {
+		klog.V(4).Infof("CSR %q was not recognized", csr.name)
+		return reconcileStop, nil
+	}
+
+	// Check if user name in csr is the same as commonName field in csr request.
+	if csr.username != commonName {
+		return reconcileContinue, nil
+	}
+
+	// Authorize whether the current spoke agent has been authorized to renew its csr.
+	allowed, err := authorize(ctx, r.kubeClient, csr)
+	if err != nil {
+		return reconcileContinue, err
+	}
+	if !allowed {
+		klog.V(4).Infof("Managed cluster csr %q cannont be auto approved due to subject access review was not approved", csr.name)
+		return reconcileStop, nil
+	}
+
+	if err := approveCSR(r.kubeClient); err != nil {
+		return reconcileContinue, err
+	}
+
+	r.eventRecorder.Eventf("ManagedClusterCSRAutoApproved", "spoke cluster csr %q is auto approved by hub csr controller", csr.name)
+	return reconcileStop, nil
+}
+
+type csrBootstrapReconciler struct {
+	kubeClient    kubernetes.Interface
+	eventRecorder events.Recorder
+	approvalUsers sets.Set[string]
+}
+
+func NewCSRBootstrapReconciler(kubeClient kubernetes.Interface, recorder events.Recorder, approvalUsers []string) Reconciler {
+	return &csrBootstrapReconciler{
+		kubeClient:    kubeClient,
+		eventRecorder: recorder.WithComponentSuffix("csr-approving-controller"),
+		approvalUsers: sets.New(approvalUsers...),
+	}
+}
+
+func (b *csrBootstrapReconciler) Reconcile(ctx context.Context, csr csrInfo, approveCSR approveCSRFunc) (reconcileState, error) {
+	// Check whether current csr is a valid spoker cluster csr.
+	if valid, _ := validateCSR(csr); !valid {
+		klog.V(4).Infof("CSR %q was not recognized", csr.name)
+		return reconcileStop, nil
+	}
+
+	// Check whether current csr can be approved.
+	if !b.approvalUsers.Has(csr.username) {
```

Comment:
do we also need to set accepthubclient=true?

Reply:
yeah, we may need it, so we will have a whole auto approval process for a spoke